metricflow/specs/query_interface.py

```diff
@@ -0,0 +1,49 @@
+from __future__ import annotations
+
+from typing import Sequence
+from abc import ABC
+
+
+class QueryInterfaceMetric(ABC):
+    """Metric in the Jinja interface."""
+
+    def __init__(self, name: str) -> None:  # noqa: D
```

Comment:
Ya, I realized that. It is up to the implementer to ensure that the same initializer signature is used each time. It is a bit unfortunate, but I couldn't think of a better way.

Reply:
I'm not sure how much of this is tied up with the anonymous class pattern elsewhere in this PR and how much is required, but if you want to have a concrete interface spec for initializing the class based on specific input parameters, you can require a class factory method. Something like:

```
@abstractmethod
@classmethod
def from_name(cls, name: str) -> Self:
    """Initialize and return an instance of this class from the name."
    raise NotImplementedError
```

Then you always invoke the equivalent of `cls.from_name(some_name)` at the callsite, and never bother with directly invoking the initializer.

This pattern also works with a Protocol type.

It might be worth it for us to have a more interactive chat - I'm realizing I can give feedback on Python mechanics but my understanding of the specific problem you're solving is limited enough that I can't really help out with the bigger questions of what we want from the type spec itself.